pallets/external-validators/src/lib.rs

```diff
@@ -0,0 +1,635 @@
+            // If more validators than max, take the first n
+            let validators = BoundedVec::truncate_from(validators);
+            <ExternalValidators<T>>::put(validators);
+
+            Ok(())
+        }
+
+        /// Helper to set a new `ForceEra` mode.
+        pub(crate) fn set_force_era(mode: Forcing) {
+            log::info!("Setting force era mode {:?}.", mode);
+            ForceEra::<T>::put(mode);
+            Self::deposit_event(Event::<T>::ForceEra { mode });
+        }
+
+        pub fn whitelisted_validators() -> Vec<T::ValidatorId> {
+            <WhitelistedValidators<T>>::get().into()
+        }
+
+        pub fn active_era() -> Option<ActiveEraInfo> {
+            <ActiveEra<T>>::get()
+        }
+
+        pub fn current_era() -> Option<EraIndex> {
+            <CurrentEra<T>>::get()
+        }
+
+        pub fn eras_start_session_index(era: EraIndex) -> Option<u32> {
+            <ErasStartSessionIndex<T>>::get(era)
+        }
+
+        /// Returns validators for the next session. Whitelisted validators first, then external validators.
+        /// The returned list is deduplicated, but the order is respected.
+        /// If `SkipExternalValidators` is true, this function will ignore external validators.
+        pub fn validators() -> Vec<T::ValidatorId> {
+            let mut validators: Vec<_> = WhitelistedValidators::<T>::get().into();
+
+            if !SkipExternalValidators::<T>::get() {
+                validators.extend(ExternalValidators::<T>::get())
+            }
+
+            remove_duplicates(validators)
+        }
+
+        /// Plan a new session potentially trigger a new era.
+        pub(crate) fn new_session(session_index: SessionIndex) -> Option<Vec<T::ValidatorId>> {
+            if let Some(current_era) = Self::current_era() {
+                // Initial era has been set.
+                let current_era_start_session_index = Self::eras_start_session_index(current_era)
+                    .unwrap_or_else(|| {
+                        frame_support::print(
+                            "Error: start_session_index must be set for current_era",
+                        );
+                        0
+                    });
+
+                let era_length = session_index.saturating_sub(current_era_start_session_index); // Must never happen.
+
+                match ForceEra::<T>::get() {
+                    // Will be set to `NotForcing` again if a new era has been triggered.
+                    Forcing::ForceNew => (),
+                    // Short circuit to `try_trigger_new_era`.
+                    Forcing::ForceAlways => (),
+                    // Only go to `try_trigger_new_era` if deadline reached.
+                    Forcing::NotForcing if era_length >= T::SessionsPerEra::get() => (),
+                    _ => {
+                        // Either `Forcing::ForceNone`,
+                        // or `Forcing::NotForcing if era_length >= T::SessionsPerEra::get()`.
```

Comment:
I think this comment is a bit confusing, as `Forcing::NotForcing if era_length >= T::SessionsPerEra::get()` is already covered by the above branch.

Reply:
It's wrong, should be `<` here (but this is copied from polkadot)

Reply:
*fixed*